### DISCUSSION TITLE
INTEGER as C-like enums ( or enums with discriminators)

### DISCUSSION THREAD

**Discussion by naveedpash** [librasn/rasn#40] [Q&A] [answered]
I have the following ASN.1

```asn.1
Result ::= INTEGER {
  acceptance (0),
  user-rejection (1),
  provider-rejection (2)
}
```

One way to represent this in rust would be 

```rust
#[derive(Debug, AsnType, Encode, Decode)]
#[rasn(automatic_tags)]
struct StructContainingResult {
  result: rasn::types::Integer
}
```

However I would then have to write code to check if `result` actually contains a 0 or 1 or 2 and return an error if it had anything else. Also my code will be less self-explanatory. Is the following a valid way to declare the same asn.1 using rasn:

``` rust
#[derive(Debug, AsnType, Encode, Decode)]
#[repr(u8)]
enum Result {
  Acceptance=0,
  UserRejection=1,
  ProviderRejection=2
}
```

(I know the ideal answer is "try and test" but I actually don't know how to calculate the encoding by hand)

**Comment by XAMPPRocky** [comment, marked as answer]
Thank you for your question. Yeah this falls under #6 as providing constraints for integers, which isn't supported directly at the moment.

For this specific instance though, you should be able to model pretty well in `rasn` by using `enumerated` and setting `#[rasn(tag(universal, 2)]`, which would encode it as an integer but also check if it matches any of the variants. It's a bit hacky, but if you don't want to write the validation it should work.

**Reply by naveedpash** [reply]
Alright for how I'll use `enumerated` along with the integer universal tag because the document I'm implementing has a lot of this kind of scenario

Wish I was literate enough to help you with constraints....
